cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,142 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of a queue. Queue allows adding elements and
+ * removing them in a predefined order (defined in subclasses).
+ *
+ * Implemented methods are not boxed, when using with some `Int` and `Long`.
+ */
+trait CQueue[@specialized(Int, Long) T] {
+  /**
+   * @return Next element to be dequeued.
+   */
+  def first(): T
+
+  /**
+   *
+   * @return Next element that is removed from the `CQueue`.
+   */
+  def deque(): T
+
+  /**
+   * Adds element to the `CQueue`.
+   */
+  def +=(elem: T)
+
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]] - elements are added to the end of an underlying list and
+ * removed from the beginning.
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  /**
+   * Adds an element at the beginning of the underlying list.
+   */
+  def enqueueFirst(elem: T)
+}
+
+object CQueue {
+  import Implicits._
```

Comment:
btw, another alternative to implicits is using typeTags. not saying it is better, but something to consider:
A toy example to do something specific for Int:

import scala.reflect.runtime.universe._
class Q[T] { }
abstract class QFac[T] {def fifo(): Q[T] }

val qFacAny = new QFac[AnyRef] { def fifo() = new Q[AnyRef] }
val qFacInt = new QFac[Int] { def fifo() = { println("return anything special for ints") ; new Q[Int] } }
object Q {
def fifo[T: TypeTag] = { if (typeTag[T] == typeTag[Int]) qFacInt.fifo() else qFacAny.fifo() }
}